agent/agent.go

```diff
@@ -62,13 +66,66 @@ var (
 	healthCheckPath      = flag.String("health-check-path", "/", "Path on backend host to issue health checks against.  Defaults to the root.")
 	healthCheckFreq      = flag.Int("health-check-interval-seconds", 0, "Wait time in seconds between health checks.  Set to zero to disable health checks.  Checks disabled by default.")
 	healthCheckUnhealthy = flag.Int("health-check-unhealthy-threshold", 2, "A so-far healthy backend will be marked unhealthy after this many consecutive failures. The minimum value is 1.")
+	sessionCookieTimeout = flag.Duration("session-cookie-timeout", 10*time.Minute, "Cookie timeout for the session cookie")
+	cookieLRU, _         = utils.NewCookieCache(cookieCacheLimit)
 )
 
 func hostProxy(ctx context.Context, host, shimPath string, injectShimCode bool) (http.Handler, error) {
 	hostProxy := httputil.NewSingleHostReverseProxy(&url.URL{
 		Scheme: "http",
 		Host:   host,
 	})
+
+	existingDirector := hostProxy.Director
+
+	hostProxy.Director = func(r *http.Request) {
+		existingDirector(r)
+
+		sessionCookie, err := r.Cookie("session-id")
+		requestDump, err := httputil.DumpRequest(r, true)
+		if err != nil {
+			log.Println(err)
+		}
+		log.Println("request in director")
+		log.Println(string(requestDump))
+
+		if err != nil {
+			// This should not happen, should have a session cookie by this point
+			log.Fatal(err)
+		} else {
+			if sessionCookie != nil {
+				sessionID := sessionCookie.Value
+				cachedCookieJar, ok := cookieLRU.GetCachedCookieJar(sessionID)
+				if !ok {
+					fmt.Printf("No cached cookie jar with session ID: %v", sessionID)
+				}
+
+				// Get only the applicable cookies
+				cookies := cachedCookieJar.(http.CookieJar).Cookies(r.URL)
+
+				// Add the cookies to the request header
+				for c := range cookies {
+					r.AddCookie(cookies[c])
+				}
+			}
+
+		}
+	}
+
+	hostProxy.ModifyResponse = func(res *http.Response) error {
+
+		res.Header.Add("Set-Cookie", "test")
```

Comment:
The format of a `Set-Cookie` header takes both a name and a value (e.g. something like `<NAME>=<VALUE>`).

However, rather than trying to set the header directly, it will be more reliable to use the [Cookie.String() method](https://golang.org/pkg/net/http/#Cookie.String).

e.g.

```go
		c := &http.Cookie{
			Name: "session-id",
			Value: sessionID,
			MaxAge: int(sessionCookieTimeout.Seconds()),
		}
		res.Header.Add("Set-Cookie", c.String())
```